gdnative-derive/src/derive_conv_variant.rs

```diff
@@ -0,0 +1,480 @@
+        .map(|bounded_ty| {
+            where_predicate(syn::Type::Path(bounded_ty), bound.clone())
+        });
+
+    let mut generics = generics.clone();
+    generics
+        .make_where_clause()
+        .predicates
+        .extend(new_predicates);
+
+    generics
+}
+
+pub(crate) fn parse_derive_input(input: TokenStream, bound: &syn::Path) -> DeriveData {
+    let input = match syn::parse_macro_input::parse::<DeriveInput>(input) {
+        Ok(val) => val,
+        Err(err) => {
+            panic!("{}", err);
+        }
+    };
+
+    let repr = match input.data {
+        Data::Struct(struct_data) => Repr::Struct(VariantRepr::repr_for(&struct_data.fields)),
+        Data::Enum(enum_data) => {
+            Repr::Enum(enum_data.variants
+                .iter()
+                .map(|variant| (variant.ident.clone(), VariantRepr::repr_for(&variant.fields)))
+                .collect())
+        },
+        Data::Union(_) => panic!("Variant conversion derive macro does not work on unions."),
+    };
+
+    let generics = extend_bounds(input.generics, &repr, bound);
+
+    DeriveData { ident: input.ident, repr, generics }
+}
+
+pub(crate) fn derive_to_variant(input: TokenStream) -> TokenStream {
+    let bound: syn::Path = syn::parse2(quote! { ::gdnative::ToVariant }).unwrap();
+    let DeriveData { ident, repr, generics } = parse_derive_input(input, &bound);
+
+    let return_expr = match repr {
+        Repr::Struct(var_repr) => {
+            let destructure_pattern = var_repr.destructure_pattern();
+            let to_variant = var_repr.to_variant();
+            quote! {
+                {
+                    let #ident #destructure_pattern = &self;
+                    #to_variant
+                }
+            }
+        },
+        Repr::Enum(variants) => {
+            let match_arms = variants
+                .iter()
+                .map(|(var_ident, var_repr)| {
+                    let destructure_pattern = var_repr.destructure_pattern();
+                    let to_variant = var_repr.to_variant();
+                    let var_ident_string = format!("{}", var_ident);
+                    let var_ident_string_literal = Literal::string(&var_ident_string);
+                    quote! {
+                        #ident::#var_ident #destructure_pattern => {
+                            let mut __dict = ::gdnative::Dictionary::new();
+                            let __key = ::gdnative::GodotString::from(#var_ident_string_literal).to_variant();
+                            let __value = #to_variant;
+                            __dict.set(&__key, &__value);
+                            __dict.to_variant()
+                        }
+                    }
+                });
+
+            quote! {
+                match &self {
+                    #( #match_arms ),*
+                }
+            }
+        },
+    };
+
+    let where_clause = &generics.where_clause;
+
+    let result = quote! {
+        #[automatically_derived]
```

Comment:
I've never seen this and can't find anything in the stdlib docs about this either, mind sharing where this attribute comes from? :smile:

Reply:
I saw it on other derived impls in `cargo-expand` while debugging, and turns out it's a currently undocumented attribute that does a couple of things, and is widely used by derive macros: https://github.com/rust-lang-nursery/reference/issues/532#issuecomment-471086171

I'm not sure if it's really needed, though. Admittedly it's mostly cargo cult here.

Reply:
Hm okay, it doesn't seem like there is much use for it, also Centril expressed concerns about destabilization or clarification of semantics, so I would probably prefer to remove this for now if it doesn't serve much purpose.

Reply:
Removed.